lib/Dialect/AIE/IR/AIEDialect.cpp

```diff
@@ -1218,6 +1218,17 @@ LogicalResult xilinx::AIE::BufferOp::verify() {
   return success();
 }
 
+// FIXME: make address assignment for buffers explicit and move this function to
+// an interface
+uint64_t xilinx::AIE::getBufferBaseAddress(Operation *bufOp) {
+  if (auto buf = dyn_cast<BufferOp>(bufOp)) {
+    return buf.address();
+  } else if (auto buf = dyn_cast<ExternalBufferOp>(bufOp)) {
+    assert(false && "External buffer addresses are assigned at runtime.");
+  }
+  llvm_unreachable("unknown buffer type");
```

Comment:
Asserts are not supposed to be used to handle error cases, but to bail on a broken/corrupted state. Here it looks like we're using it as a convenient way to "return an error". If this function can only be called on a `BufferOp`, you should assert on that. If calling on an `ExternalBufferOp` is a user error, then you should find a way to return an error (e.g.: optional), and handle it properly from the caller.

Reply:
Most place this is called are `void`. Only thing I can do here is `report_fatal_error`. I don't disagree that this is wrong but lack of correct error propagation in the codebase is too big for me to tackle by myself :shrug:.